### ISSUE TITLE
RuntimeError: Socket Timeout when setting up NCCL communicator

### ISSUE THREAD

**Issue by yulingao** [NVIDIA/Megatron-LM#386]
when i try to train gpt on HPC with slurm, I encounter this error.

pytorch 11.3, cuda 11.4, nccl 2.8(I'm not sure whether nccl has been used)

> Traceback (most recent call last):
  File "../Megatron-LM/pretrain_gpt.py", line 116, in <module>
    pretrain(train_valid_test_datasets_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/training.py", line 89, in pretrain
Traceback (most recent call last):
  File "../Megatron-LM/pretrain_gpt.py", line 116, in <module>
    pretrain(train_valid_test_datasets_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/training.py", line 89, in pretrain
Traceback (most recent call last):
  File "../Megatron-LM/pretrain_gpt.py", line 116, in <module>
    pretrain(train_valid_test_datasets_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/training.py", line 89, in pretrain
    initialize_megatron(extra_args_provider=extra_args_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 82, in initialize_megatron
    initialize_megatron(extra_args_provider=extra_args_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 82, in initialize_megatron
    initialize_megatron(extra_args_provider=extra_args_provider,
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 82, in initialize_megatron
    _compile_dependencies()
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 133, in _compile_dependencies
    torch.distributed.barrier()
  File "/gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2776, in barrier
    _compile_dependencies()
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 133, in _compile_dependencies
    torch.distributed.barrier()
  File "/gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2776, in barrier
    _compile_dependencies()
  File "/gs/home/momo/pytorch/transformer/Megatron-LM/megatron/initialize.py", line 133, in _compile_dependencies
    torch.distributed.barrier()
  File "/gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2776, in barrier
    work = default_pg.barrier(opts=opts)
RuntimeError: [2] is setting up NCCL communicator and retreiving ncclUniqueId from [0] via c10d key-value store by key '0', but store->get('0') got error: Socket Timeout
Exception raised from recvBytes at ../torch/csrc/distributed/c10d/Utils.hpp:580 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x42 (0x7f6fb68207d2 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x5f (0x7f6fb681cf3f in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #2: c10d::TCPStore::doWait(c10::ArrayRef<std::string>, std::chrono::duration<long, std::ratio<1l, 1000l> >) + 0x11f (0x7f6ff2e052ff in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #3: c10d::TCPStore::doGet(std::string const&) + 0x21 (0x7f6ff2e06281 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10d::TCPStore::get(std::string const&) + 0x5b (0x7f6ff2e0630b in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #5: c10d::PrefixStore::get(std::string const&) + 0x32 (0x7f6ff2dd7e52 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #6: c10d::PrefixStore::get(std::string const&) + 0x32 (0x7f6ff2dd7e52 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #7: c10d::PrefixStore::get(std::string const&) + 0x32 (0x7f6ff2dd7e52 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #8: c10d::ProcessGroupNCCL::broadcastUniqueNCCLID(ncclUniqueId*, c10d::OpType, std::string const&, int) + 0xe4 (0x7f70098ba4b4 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #9: c10d::ProcessGroupNCCL::getNCCLComm(std::string const&, std::vector<c10::Device, std::allocator<c10::Device> > const&, c10d::OpType, int, bool) + 0x1d9 (0x7f70098be549 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #10: <unknown function> + 0x26ff9e5 (0x7f70098c19e5 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #11: c10d::ProcessGroupNCCL::allreduce_impl(std::vector<at::Tensor, std::allocator<at::Tensor> >&, c10d::AllreduceOptions const&) + 0xf (0x7f70098c2cdf in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #12: c10d::ProcessGroupNCCL::allreduce(std::vector<at::Tensor, std::allocator<at::Tensor> >&, c10d::AllreduceOptions const&) + 0x2d3 (0x7f70098c8df3 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #13: c10d::ProcessGroupNCCL::barrier(c10d::BarrierOptions const&) + 0x72a (0x7f70098d284a in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so)
frame #14: <unknown function> + 0x7fda43 (0x7f706c135a43 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
frame #15: <unknown function> + 0x1f8132 (0x7f706bb30132 in /gs/home/momo/anaconda3/envs/cuda11.4/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
frame #16: PyCFunction_Call + 0x52 (0x4f5652 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #17: _PyObject_MakeTpCall + 0x3bb (0x4e0c8b in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #18: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x4f53fd]
frame #19: _PyEval_EvalFrameDefault + 0x1150 (0x4d9140 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #20: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x586b70]
frame #21: _PyEval_EvalFrameDefault + 0x49a9 (0x4dc999 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #22: _PyFunction_Vectorcall + 0x106 (0x4e7fe6 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #23: _PyEval_EvalFrameDefault + 0x399 (0x4d8389 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #24: _PyEval_EvalCodeWithName + 0x2f1 (0x4d6fb1 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #25: _PyFunction_Vectorcall + 0x19c (0x4e807c in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #26: _PyEval_EvalFrameDefault + 0x1150 (0x4d9140 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #27: _PyEval_EvalCodeWithName + 0x2f1 (0x4d6fb1 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #28: _PyFunction_Vectorcall + 0x19c (0x4e807c in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #29: _PyEval_EvalFrameDefault + 0x1150 (0x4d9140 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #30: _PyEval_EvalCodeWithName + 0x2f1 (0x4d6fb1 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #31: PyEval_EvalCodeEx + 0x39 (0x585d79 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #32: PyEval_EvalCode + 0x1b (0x585d3b in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #33: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x5a5a91]
frame #34: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x5a4a9f]
frame #35: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x45c417]
frame #36: PyRun_SimpleFileExFlags + 0x340 (0x45bfb8 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #37: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x44fd9e]
frame #38: Py_BytesMain + 0x39 (0x579dd9 in /gs/home/momo/anaconda3/envs/cuda11.4/bin/python)
frame #39: __libc_start_main + 0xf5 (0x7f70741f7a85 in /lib64/libc.so.6)
frame #40: /gs/home/momo/anaconda3/envs/cuda11.4/bin/python() [0x579c8d]

**Comment by jon-barker** [comment]
Are you using Docker on slurm? If so, do you have a Dockerfile or container I could use to reproduce the error?

**Comment by yulingao** [comment]
> Are you using Docker on slurm? If so, do you have a Dockerfile or container I could use to reproduce the error?

I want to use docker, but no docker or container in this HPC that i'm using, only provide anaconda3, and have no access to the internet.

**Comment by jon-barker** [comment]
OK, can you share the configuration of the training job then? i.e. the arguments to `pretrain_gpt.py`. Thanks

**Comment by yulingao** [comment]
of course, here is my configuration file, thank you for your time

> #!/bin/bash
#SBATCH --job-name=train_gpt2
#SBATCH --partition=gpu01
#SBATCH --nodes=1
#SBATCH --gres=gpu:4
#SBATCH --time=12:00:00
#SBATCH -o out.std
#SBATCH -e out.err

conda activate cuda11.4

cd $SLURM_SUBMIT_DIR

echo "SLURM_JOBID= "$SLURM_JOBID
echo "SLURM_JOB_NODELIST= "$SLURM_JOB_NODELIST

GPUS_PER_NODE=4
MASTER_ADDR=localhost
MASTER_PORT=6001
NNODES=1
NODE_RANK=0
WORLD_SIZE=$(($GPUS_PER_NODE*$NNODES))

DATA_PATH=data/meg-gpt2_text_document
CHECKPOINT_PATH=./ck

export CUDA_DEVICE_MAX_CONNECTIONS="1"

DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT"

python -m torch.distributed.launch $DISTRIBUTED_ARGS \
       ../Megatron-LM/pretrain_gpt.py \
       --tensor-model-parallel-size 1 \
       --pipeline-model-parallel-size 1 \
       --num-layers 12 \
       --hidden-size 768 \
       --num-attention-heads 12 \
       --micro-batch-size 4 \
       --global-batch-size 16 \
       --seq-length 1024 \
       --max-position-embeddings 1024 \
       --train-iters 5000 \
       --lr-decay-iters 320000 \
       --save $CHECKPOINT_PATH \
       --load $CHECKPOINT_PATH \
       --data-path $DATA_PATH \
       --vocab-file data/vocab.json \
       --merge-file data/merges.txt \
       --data-impl mmap \
       --split 949,50,1 \
       --distributed-backend nccl \
       --lr 0.00015 \
       --lr-decay-style cosine \
       --min-lr 1.0e-5 \
       --weight-decay 1e-2 \
       --clip-grad 1.0 \
       --lr-warmup-fraction .01 \
       --log-interval 10 \
       --save-interval 500 \
       --eval-interval 100 \
       --eval-iters 10 \
       --fp16

echo "Job end"

**Comment by jon-barker** [comment]
I haven't been able to replicate this result yet, but will try again with a container tomorrow with those exact versions of torch, cuda and nccl. One thing you might try is upgrading your nccl as it's quite old and the source of the error appears to be nccl.

**Comment by plegresl** [comment]
Can you please try re-running it with the line "export NCCL_DEBUG=INFO" somewhere above the line that starts with "python -m torch.distributed.launch ..."?

(see https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-debug for more information about what this does)

**Comment by yulingao** [comment]
> I haven't been able to replicate this result yet, but will try again with a container tomorrow with those exact versions of torch, cuda and nccl. One thing you might try is upgrading your nccl as it's quite old and the source of the error appears to be nccl.

Sad thing. I can't upgrade nccl as I don't have administrative rights for the HPC.

**Comment by yulingao** [comment]
> Can you please try re-running it with the line "export NCCL_DEBUG=INFO" somewhere above the line that starts with "python -m torch.distributed.launch ..."?
> 
> (see https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-debug for more information about what this does)

I have tried to add the following to slurm sbatch scripts. But still no nccl related output in out.err or out.std. Do you think it could be that i didn't call nccl correctly?

export NCCL_SOCKET_IFNAME=ens4f0
export NCCL_DEBUG=INFO
export NCCL_DEBUG_SUBSYS=ALL

**Comment by plegresl** [comment]
Unless you're absolutely certain about the correct interface I would recommend **_not_** setting NCCL_SOCKET_IFNAME. Instead of trying to debug this with Megatron-LM can you try running this simple test program:

`import os
import torch
import torch.distributed as dist

dist.init_process_group(backend='nccl')

rank = os.getenv('RANK', -1)

print(rank, 'entering barrier...')
torch.distributed.barrier()
print(rank, 'through barrier...')`

**Comment by yulingao** [comment]
> Unless you're absolutely certain about the correct interface I would recommend **_not_** setting NCCL_SOCKET_IFNAME. Instead of trying to debug this with Megatron-LM can you try running this simple test program:
> 
> `import os import torch import torch.distributed as dist
> 
> dist.init_process_group(backend='nccl')
> 
> rank = os.getenv('RANK', -1)
> 
> print(rank, 'entering barrier...') torch.distributed.barrier() print(rank, 'through barrier...')`

I save this test program into test.py, run it with "python -m torch.distributed.launch test.py", and get the following result. It seems like that NCCL works well.

> 0 entering barrier...
0 through barrier...
FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use_env is set by default in torchrun.
If your script expects `--local_rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

**Comment by yulingao** [comment]
Also, this error occurs when "compiling and loading fused kernels", that is, lines  133 of megatron/initialize.py. 

I wonder if the socket timeout is caused by fused_kernels.load(args) being stuck or not loading within the default 10 minutes?

    # Always build on rank zero first.
    if torch.distributed.get_rank() == 0:
        start_time = time.time()
        print('> compiling and loading fused kernels ...', flush=True)
        fused_kernels.load(args)
        torch.distributed.barrier()
    else:
        torch.distributed.barrier()   # this is line 133
        fused_kernels.load(args)
    # Simple barrier to make sure all ranks have passed the
    # compilation phase successfully before moving on to the
    # rest of the program. We think this might ensure that
    # the lock is released.
    torch.distributed.barrier()

**Comment by yulingao** [comment]
I try to add  the load time for "fused_kernel.load(args)" by specifying  the " --distributed-timeout-minutes 30 " in the training args. It's still reporting the same "socket timeout" error, so I'm basically sure it's this "fused_kernel.load(args)" that's stuck.

**Comment by yulingao** [comment]
@jon-barker @plegresl 
Hi, I find the problem. The first time I run Megatron-LM, the "torch.utils.cpp_extension.load" found the wrong cuda (10.1) because the CUDA_HOME environment was not configured, so the fused_kernel did not compile successfully(that is, scaled_masked_softmax_cuda.so, scaled_softmax_cuda.so, scaled_upper_triang_masked_softmax_cuda.so). "torch.utils.cpp_extension.load" left a lock file in the /path/to/Megatron-LM/megatron/fused_kernels/build directory, which prevented subsequent compilation of the program at runtime, resulting in a socket timeout. 

I have now removed this lock file, and specified the CUDA_HOME in the slurm submit file, I was able to start training properly.

Thank you so much for your help!
